// tap/lib/tap-runner.js
var fs = require("fs")
  , child_process = require("child_process")
  , path = require("path")
  , chain = require("slide").chain
  , asyncMap = require("slide").asyncMap
  , TapProducer = require("./tap-producer.js")
  , TapConsumer = require("./tap-consumer.js")
  , assert = require("./tap-assert.js")
  , inherits = require("inherits")
  , util = require("util")
  , CovHtml = require("./tap-cov-html.js")
  , doCoverage = process.env.TAP_COV
               || process.env.npm_package_config_coverage
               || process.env.npm_config_coverage

var Runner = module.exports = function (dir, diag, cb) {
  Runner.super.call(this, diag)

  this.doCoverage = doCoverage
  // An array of full paths to files to obtain coverage
  this.coverageFiles = []
  // The source of these files
  this.coverageFilesSource = {}
  // Where to write coverage information
  this.coverageOutDir = "./coverage"
  // Temporary test files bunkerified we'll remove later
  this.f2delete = []
  // Raw coverage stats, as read from JSON files
  this.rawCovStats = []
  // Processed coverage information, per file to cover:
  this.covStats = {}

  if (dir) {
    var filesToCover = "./lib"
      , coverageOutDir = "./coverage"

    if (doCoverage) {
      dir = dir.filter(function(arg) {
        if (arg.match(/^--cover=/)) {
          filesToCover = arg.split("--cover=")[1]
          return false
        } else if (arg.match(/^--cover-dir=/)) {
          coverageOutDir = arg.split("--cover-dir=")[1]
          return false
        }
        return
